# Questions 8-10

In [37]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [10]:
house_price_final = pd.read_excel("C:/Users/SNIGGDHA/SA2/SA2_Q8-10.xlsx")

print(house_price_final)

         Id   price  sqft_living  sqft_living * Expensive  bedrooms 4-6   \
0      3125  269100         2190                        0              0   
1     10559  445000         1990                        0              0   
2      8819  400000         1730                        0              0   
3      1197  215000         1900                        0              1   
4     16507  657500         4140                        0              1   
...     ...     ...          ...                      ...            ...   
4995   4696  305000         1320                        0              0   
4996   6551  348000         1670                        0              0   
4997  11488  470000         2100                        0              0   
4998    959  206600         1390                        0              0   
4999  10556  445000         2250                        0              0   

      bedrooms 7 and more  bathrooms 0-3  floors 1-2  floors 3 and more  \
0           

## Question 8

In [8]:
import statsmodels.api as sm
from statsmodels.compat import lzip
from statsmodels.stats.diagnostic import het_breuschpagan


In [12]:
house_price_final.columns

Index(['Id', 'price', 'sqft_living', 'sqft_living * Expensive',
       'bedrooms 4-6 ', 'bedrooms 7 and more', 'bathrooms 0-3', 'floors 1-2',
       'floors 3 and more', 'basement_present (YES = 1)',
       'waterfront (YES = 1)', 'Distance', 'age'],
      dtype='object')

In [15]:
# dependent variable
y = house_price_final['price']

# independent variables
X=house_price_final[['sqft_living', 'bedrooms 4-6 ', 'bedrooms 7 and more', 'bathrooms 0-3', 'floors 1-2', 'floors 3 and more', 'basement_present (YES = 1)', 'waterfront (YES = 1)', 'Distance', 'age']]

sqft_living  bedrooms 4-6   bedrooms 7 and more  bathrooms 0-3  \
0            2190              0                    0              1   
1            1990              0                    0              1   
2            1730              0                    0              1   
3            1900              1                    0              1   
4            4140              1                    0              1   
...           ...            ...                  ...            ...   
4995         1320              0                    0              1   
4996         1670              0                    0              1   
4997         2100              0                    0              1   
4998         1390              0                    0              1   
4999         2250              0                    0              1   

      floors 1-2  floors 3 and more  basement_present (YES = 1)  \
0              1                  0                           0   
1              1                  0                           0   
2              1                  0                           0   
3              1                  0                           0   
4              1                  0                           0   
...          ...                ...                         ...   
4995           1                  0                           0   
4996           1                  0                           0   
4997           1                  0                           0   
4998           1                  0                           0   
4999           1                  0                           0   

      waterfront (YES = 1)   Distance  age  
0                        0  24.422199   20  
1                        0  13.876083   30  
2                        0   7.615336   53  
3                        0  14.188716   47  
4                        0  22.370637   12  
...                    ...        ...  ...  
4995                     0   5.314980   10  
4996                     0  11.250505   12  
4997                     0  22.332847   15  
4998                     0  18.649987   27  
4999                     0  26.145412   21  

[5000 rows x 10 columns]

In [66]:
# Adding a constant term to the independent variables
X = sm.add_constant(X)

# Fitting the ordinary least squares (OLS) model
model = sm.OLS(y, X)
results = model.fit()

# Calculating residuals
residuals = results.resid

# Performing the Breusch-Pagan test
# Ho: Data shows no Heteroskedasticity 
# Ha: Data shows Heteroskedasticity 

test_stats, p_value, fvalue, f_p_value = het_breuschpagan(residuals, X)

if p_value < 0.05: 
    print("Reject the null hypothesis as there is Heteroskedasticity in the dataset \n") 
else: 
    print("Fail to reject the null hypothesis as there is no Heteroskedasticity in the dataset \n") 

# Print the test result
print("Breusch-Pagan Test:")
print("Test statistic:", test_stats)
print("p-value:", p_value)

Reject the null hypothesis as there is Heteroskedasticity in the dataset 

Breusch-Pagan Test:
Test statistic: 977.7580563009935
p-value: 1.1552894099424838e-203


## Question 9

In [51]:
# Ho: Residuals are Normally Distributed
# Ha: Residuals are not Normally Distributed
# We will use Jarque Bera Test of Normality

teststats, pvalue = stats.jarque_bera(residuals)

print("Jarque Bera Test:")
print("Test statistic:", teststats)
print("p-value:", pvalue)

if pvalue < 0.05: 
    print("\nReject the null hypothesis as residuals are not Normally Distributed using Jarque Bera Test of Normality \n") 
else: 
    print("\nFail to reject the null hypothesis as residuals are not Normally Distributed using Jarque Bera Test of Normality \n") 


Jarque Bera Test:
Test statistic: 35879.68670559837
p-value: 0.0

Reject the null hypothesis as residuals are not Normally Distributed using Jarque Bera Test of Normality 



In [54]:
# Ho: Residuals are Normally Distributed
# Ha: Residuals are not Normally Distributed
# We will use Shapiro-Wilk Test of Normality
teststats, pvalue = stats.shapiro(residuals)


print("Shapiro-Wilk Test:")
print("Test statistic:", teststats)
print("p-value:", pvalue)


if pvalue < 0.05: 
    print("\nReject the null hypothesis as residuals are not Normally Distributed using Shapiro-Wilk Test of Normality \n") 
else: 
    print("\nFail to reject the null hypothesis as residuals are not Normally Distributed using Shapiro-Wilk Test of Normality \n") 


Shapiro-Wilk Test:
Test statistic: 0.8786818981170654
p-value: 0.0

Reject the null hypothesis as residuals are not Normally Distributed using Shapiro-Wilk Test of Normality 



## Question 10

In [56]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.663
Model:                            OLS   Adj. R-squared:                  0.663
Method:                 Least Squares   F-statistic:                     983.4
Date:                Sun, 21 May 2023   Prob (F-statistic):               0.00
Time:                        11:40:37   Log-Likelihood:                -68276.
No. Observations:                5000   AIC:                         1.366e+05
Df Residuals:                    4989   BIC:                         1.366e+05
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================================
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
const                       5.281e+05   4.28e+04     12.351      0.000    4.44e+05    6.12e+05
sqft_living                  281.5012      4.637     60.712      0.000     272.411     290.591
bedrooms 4-6               -5.961e+04   7105.810     -8.390      0.000   -7.35e+04   -4.57e+04
bedrooms 7 and more        -3.542e+05   4.96e+04     -7.142      0.000   -4.51e+05   -2.57e+05
bathrooms 0-3              -1.239e+05   1.25e+04     -9.888      0.000   -1.49e+05   -9.94e+04
floors 1-2                  -2.51e+05    3.7e+04     -6.785      0.000   -3.24e+05   -1.78e+05
floors 3 and more          -2.403e+05   4.15e+04     -5.783      0.000   -3.22e+05   -1.59e+05
basement_present (YES = 1) -5.204e+04   6510.202     -7.994      0.000   -6.48e+04   -3.93e+04
waterfront (YES = 1)        7.591e+05   3.44e+04     22.071      0.000    6.92e+05    8.27e+05
Distance                    -1.79e+04    519.717    -34.450      0.000   -1.89e+04   -1.69e+04
age                          729.3900    118.379      6.161      0.000     497.314     961.465
==============================================================================
Omnibus:                     2067.911   Durbin-Watson:                   1.997
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            35879.687
Skew:                           1.523   Prob(JB):                         0.00
Kurtosis:                      15.765   Cond. No.                     5.23e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.23e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [57]:
results.params

const                         528061.132324
sqft_living                      281.501191
bedrooms 4-6                  -59614.620051
bedrooms 7 and more          -354167.067042
bathrooms 0-3                -123943.363995
floors 1-2                   -251030.683455
floors 3 and more            -240274.328585
basement_present (YES = 1)    -52041.641164
waterfront (YES = 1)          759117.068708
Distance                      -17904.381427
age                              729.389962
dtype: float64

In [58]:
house_price_final.columns

Index(['Id', 'price', 'sqft_living', 'sqft_living * Expensive',
       'bedrooms 4-6 ', 'bedrooms 7 and more', 'bathrooms 0-3', 'floors 1-2',
       'floors 3 and more', 'basement_present (YES = 1)',
       'waterfront (YES = 1)', 'Distance', 'age'],
      dtype='object')

## Regression Equation:

price = 528,061.13 + (281.50 * sqft_living) +( - 59,614.62 * bedrooms 4-6) + ( - 354,167.07 * bedrooms 7 and more ) + ( - 123,943.36 * bathrooms 0-3) + ( - 251,030.68 * floors 1-2 ) + ( - 240,274.33 * floors 3 and more ) + ( - 52,041.64 * basement_present (YES = 1) ) + (759,117.07 * waterfront (YES = 1)) + (- 17,904.38 * Distance) + (729.39 * age)
